Added new parameter

derNiklaas opened the pull request:
Closes #54 

Added:
- [x] Integer parameter
- [x] Double parameter
- [x] Boolean parameter
- [x] List parameter
- [x] Map parameter
- [x] Color parameter
- [x] URI / URL parameter
- [ ] Time parameter

API Pullrequest: https://github.com/codeoverflow-org/chatoverflow-api/pull/18

justgerd commented on src/main/scala/org/codeoverflow/chatoverflow/requirement/parameter/ListParameterImpl.scala:
```diff
@@ -0,0 +1,50 @@
+        out = out + "," + part
+      }
+    }
+    out
```
what about `value.asScala.mkString(",")` instead of this abomination?
(plus an import of `scala.collection.JavaConverters._`)

justgerd commented on src/main/scala/org/codeoverflow/chatoverflow/requirement/parameter/ListParameterImpl.scala:
```diff
@@ -0,0 +1,50 @@
+      for (part <- splits) {
+        list += part
+      }
+      set(mutableSeqAsJavaList[String](list))
```
again, complete overkill in my opinion, `value.split(",").asJava` with import in the previous comment should suffice.

justgerd commented on src/main/scala/org/codeoverflow/chatoverflow/requirement/parameter/MapParameterImpl.scala:
```diff
@@ -0,0 +1,58 @@
+        out = out + ",("+key+";"+valueSet+")"
+      }
+    })
+    out
```
import of `scala.collection.JavaConverters._`
```
value.asScala.map({ case (key, value) => s"($key;$value)" }).mkString(",")
```

justgerd commented on src/main/scala/org/codeoverflow/chatoverflow/requirement/parameter/MapParameterImpl.scala:
```diff
@@ -0,0 +1,58 @@
+          map += (parts(0) -> parts(1))
+        }
+      }
+      if(map.nonEmpty) set(mutableMapAsJavaMap[String, String](map))
```
oh my god what the fuck.
import above as usual
```
val kvpair = r"\((.+);(.+)\)".r
val konly = r"\((.+);\)".r
value.split(",")
     .map( { case kvpair(k,v) => (k -> v); case konly(k) => (k -> ""); case _ => ("" -> "") } )
     .toMap.asJava
```
yes, this is also an abomination, but slightly better in my opinion as it's functional-ish

J0B10 commented on src/main/scala/org/codeoverflow/chatoverflow/requirement/parameter/ColorParameterImpl.scala:
```diff
@@ -0,0 +1,53 @@
+    val int3 = "^(\\d+),(\\d+),(\\d+)$".r
+    val int4 = "^(\\d+),(\\d+),(\\d+),(\\d+)$".r
+    val float3 = "^(\\d+(?:\\.\\d+)?),(\\d+(?:\\.\\d+)?),(\\d+(?:\\.\\d+)?)$".r
+    val float4 = "^(\\d+(?:\\.\\d+)?),(\\d+(?:\\.\\d+)?),(\\d+(?:\\.\\d+)?),(\\d+(?:\\.\\d+)?)$".r
```
It also would be nice to have some comments here that help to understand which regex is for which syntax.

derNiklaas replied:
the name of the regex already says that:
int3 -> 3 ints
float4 -> 4 floats

derNiklaas commented:
@joblo2213 Hi there. Thanks for the feedback.
I added a URI parameter (can be converted to URL by using URI.toURL() ). also converted java map/list to scala map/list. Also added comments to the regex.

Could you do the time parameter?

hlxid commented on src/main/scala/org/codeoverflow/chatoverflow/requirement/parameter/ColorParameterImpl.scala:
```diff
@@ -0,0 +1,53 @@
+        case float4(r, g, b, a) => set(new Color(r.toFloat, g.toFloat, b.toFloat, a.toFloat))
+      }
+    } catch {
+      case _: Exception => print("Can't convert String to color")
```
Please use the logger for all catch blocks by first importing it and then extend your class with `WithLogger`
```scala
import org.codeoverflow.chatoverflow.WithLogger

class ColorParameterImpl extends ColorParameter with WithLogger {
```
and then use the logger to log failures:
```suggestion
      case _: Exception => logger error "Can't convert String to color"
```

hlxid replied:
Not necessary anymore as exceptions should be thrown.

hlxid commented on src/main/scala/org/codeoverflow/chatoverflow/requirement/parameter/UriParameterImpl.scala:
```diff
@@ -0,0 +1,20 @@
+
+  override def get(): URI = value
+
+  override def deserialize(value: String): Unit = set(new URI(value))
```
Please add exception handling here for `URISyntaxException` that may be thrown by the `URI` constructor.

J0B10 replied:
I think thorwing this exception uncought is intended behaviour here. 
Any occuring exceptions are cought by the ConfigurationService and will inform the framework that the Parameter couldn't be set.

hlxid replied:
After looking into it, throwing exceptions is the correct behavior, the rest api also reports to the gui that it has failed to set the requirement if a exception is thrown. 

This means that all error handling of all other parameters should be removed.

J0B10 commented:
After commit aacdd21 and the changes @daniel0611 requested I think there is some confusion about how exception handling should work for the parameters:

I think it should be the following way: 
Any exceptions that are thrown in `deserialize()` should get passed.

https://github.com/derNiklaas/chatoverflow/blob/e5bbc816311c92d48d59caa50a974bc9ea14140b/src/main/scala/org/codeoverflow/chatoverflow/configuration/ConfigurationService.scala#L290-L301:
```java
            try {
              val reqContent = loadedRequirementType.get.newInstance().asInstanceOf[io.Serializable]
              requirements.getAccess.setRequirementContent(requirementId, reqContent)
              reqContent.deserialize(content)


              logger info s"Created requirement content for '$requirementId' and deserialized its content."
              true
            } catch {
              case e: Exception =>
                logger error s"Unable to instantiate requirement content. Exception: ${e.getMessage}"
                false
            }
```

Here all exceptions that are thown while deserializing are cought and logged, and the framework keeps the information that not all requirements could be fullfilled so that it could also be displayed to the gui.

Thats why form my point of view no messages need to be logged and all exceptions may be uncought.
Also if the deserialization failed (like in the ColorPrameter) an exception should be thrown.

The only thing we may think about is making sure that the message of the exception does tell the user what actually is wrong.

derNiklaas commented:
Just removed Error handlers from Parameter classes.

derNiklaas commented:
https://github.com/codeoverflow-org/chatoverflow/pull/148/commits/895f22f45bc1fb8bbb99ab89cd0353f8a3a8d6e0 fixed an bug which was introduced by https://github.com/codeoverflow-org/chatoverflow/pull/148/commits/b93b8592a28614e9e9e57a375f37227fe03da8a2 . 

`(test;test2),(test3;test4),(test5,test6)` would turn into `(test;test2),(test3;test4),(;)` which shouldn't happen.
it now throws an IllegalArgumentException.
